src/kernel/bpf/prog.rs

```diff
@@ -154,12 +164,103 @@ impl BpfInfoCache {
 }
 
 
+#[derive(Debug)]
+struct LineInfoRecord {
+    path: Rc<Path>,
+    line: u32,
+    col: u16,
+}
+
+
+/// Query BPF program line information.
+fn query_line_info(
+    bpf_fd: BorrowedFd<'_>,
+    info: &sys::bpf_prog_info,
+) -> Result<Option<HashMap<Addr, LineInfoRecord>>> {
+    let prog_id = info.id;
+
+    assert_eq!(
+        info.line_info_rec_size,
+        size_of::<sys::bpf_line_info>() as _
+    );
+    let mut line_info = Vec::<sys::bpf_line_info>::with_capacity(info.nr_line_info as _);
+    // SAFETY: `bpf_line_info` is valid for any bit pattern, so we
+    //         can adjust the vector's length to its capacity.
+    let () = unsafe { line_info.set_len(line_info.capacity()) };
+
+    assert_eq!(info.jited_line_info_rec_size, size_of::<u64>() as _);
+    let mut jited_line_info = Vec::<u64>::with_capacity(info.nr_jited_line_info as _);
+    // SAFETY: `u64` is valid for any bit pattern, so we can adjust
+    //         the vector's length to its capacity.
+    let () = unsafe { jited_line_info.set_len(jited_line_info.capacity()) };
+
+    let mut info = sys::bpf_prog_info {
+        nr_line_info: info.nr_line_info,
+        line_info_rec_size: info.line_info_rec_size,
+        line_info: line_info.as_mut_ptr() as _,
+        nr_jited_line_info: info.nr_jited_line_info,
+        jited_line_info_rec_size: info.jited_line_info_rec_size,
+        jited_line_info: jited_line_info.as_mut_ptr() as _,
+        ..Default::default()
+    };
+    let () = sys::bpf_prog_get_info_from_fd(bpf_fd.as_raw_fd(), &mut info).with_context(|| {
+        format!("failed to retrieve BPF program information for program {prog_id}")
+    })?;
+
+    let mut line_records = HashMap::with_capacity(info.nr_jited_line_info as _);
```

Comment:
I think you misinterpreting line records. They are ranges, not points. See https://github.com/torvalds/linux/blob/master/kernel/bpf/log.c#L332-L371 for how kernel queries it, it finds the closest match and then it's assumed that this line info applies to all instructions up to next line_info.

So, you don't need hashmap, it will be returned as an already sorted array, so you'll be doing a binary search (but not the exact match, closest match, one of those lower/upper bound kinds)

Reply:
That makes more sense, will adjust. Allow me to question the "already sorted" aspect, though.

```
kernel reported JIT information is not sorted: [
    0xffffffffc0058c38,
    0xffffffffc0058c50,
    0xffffffffc0058c69,
    0xffffffffc0058c6e,
    0xffffffffc0058c89,
    0xffffffffc0058c95,
    0xffffffffc0058c99,
    0xffffffffc0058c9d,
    0xffffffffc0058cb4,
    0xffffffffc0058ccc,
    0xffffffffc0058cd6,
    0xffffffffc0058cea,
    0xffffffffc0045a4c,
    0xffffffffc0045a69,
    0xffffffffc0045a70,
    0xffffffffc0045a74,
    0xffffffffc0045a7e,
    0xffffffffc0045a82,
]
```